pkg/isoeditor/kargs_test.go

```diff
@@ -0,0 +1,134 @@
+      "offset": 1870,
+      "path": "isolinux/isolinux.cfg"
+    }
+  ],
+  "size": 1137
+}
+`
+		grubFileWithEmbedArea = `
+function load_video {
+  insmod efi_gop
+}
+
+insmod ext2
+
+set timeout=5
+### END /etc/grub.d/00_header ###
+
+### BEGIN /etc/grub.d/10_linux ###
+menuentry 'Fedora CoreOS (Live)' --class fedora --class gnu-linux --class gnu --class os {
+	linux /images/pxeboot/vmlinuz mitigations=auto,nosmt coreos.liveiso=fedora-coreos-35.20220103.3.0 ignition.firstboot ignition.platform.id=metal
+################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################################ COREOS_KARG_EMBED_AREA
+	initrd /images/pxeboot/initrd.img /images/ignition.img
+}
+`
+		grubFileWithoutEmbedArea = `
+function load_video {
+  insmod efi_gop
+}
+
+insmod ext2
+
+set timeout=5
+### END /etc/grub.d/00_header ###
+
+### BEGIN /etc/grub.d/10_linux ###
+menuentry 'Fedora CoreOS (Live)' --class fedora --class gnu-linux --class gnu --class os {
+	linux /images/pxeboot/vmlinuz mitigations=auto,nosmt coreos.liveiso=fedora-coreos-35.20220103.3.0 ignition.firstboot ignition.platform.id=metal
+	initrd /images/pxeboot/initrd.img /images/ignition.img
+}
+`
+	)
+
+	mockFileReaderCreator := func(ret []byte, err error) FileReader {
+		return func(_, _ string) ([]byte, error) {
+			return ret, err
+		}
+	}
+	mockFileReaderSuccess := func(fileData string) FileReader {
+		return mockFileReaderCreator([]byte(fileData), nil)
+	}
+	mockFileReaderFailure := func() FileReader {
+		return mockFileReaderCreator(nil, errors.New("this is an error"))
+	}
+
+	mockBoundariesFinderCreator := func(start, length int64, err error) BoundariesFinder {
+		return func(_, _ string) (int64, int64, error) {
+			return start, length, err
+		}
+	}
+
+	mockBoundariesFinderSuccess := func(start, length int64) BoundariesFinder {
+		return mockBoundariesFinderCreator(start, length, nil)
+	}
+
+	mockBoundariesFinderFailure := func() BoundariesFinder {
+		return mockBoundariesFinderCreator(0, 0, errors.New("this is an error"))
+	}
+
+	Context("kargs files", func() {
```

Comment:
I'm going to reply to your old comment here just so it's not outdated, but TLDR is this should be `Describe("kargsFiles"` because it's grouping tests of that function. It's good to have the function name right in the test.

---------------------------------------------------------------------------------------
In reply to https://github.com/openshift/assisted-image-service/pull/99#discussion_r1005259247:

> BTW, go does not allow reference from non-test code to test code.

This is interesting. I would expect this if the tests were in a different package, but it's surprising to me that it seems to omit `*_test.go` files when compiling main.
That said, if you reference things defined in test files from real ones and run the tests, it works.
To me it's unintuitive and feels like "magic" so I'd like to avoid defining stuff at the top level in tests anyway.

> I think `Describe` is unexpected within `Context` although it is allowed.

I don't think so at all. `Describe` and `Context` are functionally identical. They only differ semantically. Generally I use `Describe` to group tests of a particular function and `Context` when I need shared setup for tests of multiple functions.
If you need shared setup only for the tests of a single function, I'd put it in that function's `Describe` block.

https://github.com/openshift/assisted-image-service/blob/d7fecdd94b334150af7daf53b9fca04722d32582/pkg/isoeditor/isoutil_test.go is a good example where multiple `Describe`s need the same context.

> The text with karg helpers looks unclear to me.

The exact text isn't super important in this case since every test is in the same context. It might as well stay `"kargs tests"`

TLDR is `Describe` is for grouping `It` blocks for the same function and `Context` is for denoting shared setup/teardown.

Reply:
> That said, if you reference things defined in test files from real ones and run the tests, it works.

No it doesn't

I usually use `Describe` as the most external function.  That's the most common use in the assisted service.  There are many few external `Context` functions.

If you think it is very significant I will change.

Reply:
>No it doesn't

I wouldn't have asserted something like this if I didn't test it :wink: 
I have a module written that reproduces this behavior, but that said I did also find https://github.com/golang/go/issues/10184#issuecomment-84465873 which explains how it could never cause a _real_ problem.

> If you think it is very significant I will change.

Not _very_ significant, but I wrote this all up mostly to just explain how I think about writing tests. If my descriptions and examples haven't impacted how you want to write these up that's okay. As I said `Context` and `Describe` are functionally equivalent.

The only thing I would very much like to see is the container (`Context` or `Describe`) that is grouping tests for a particular function be named after that function. That way it is very clear what is being tested when a test fails.

Reply:
> I wouldn't have asserted something like this if I didn't test it wink
I have a module written that reproduces this behavior, but that said I did also find https://github.com/golang/go/issues/10184#issuecomment-84465873 which explains how it could never cause a real problem.

Well that's strange because I tested it too.
In the same package, I tried to reference in non-test file a `const` that is defined in test file.  The symbol was invisible.
Specifically tried on the `const` you made the initial comment on.